feat(platforms): Add ARMv6 builds

wtaylor opened the pull request:
Add ARMv6 builds using the work started in PR #18514.

Extends CI pipeline for the new target architecture to publish to docker and s3. Docker build script updated to a "best endeavours" approach where it will filter out bases that don't support the target architecture, this was required as distroless does not publish arm v6 binaries.

fixes #18445

jszwedko committed:
feat(platforms): Add ARMv6 builds

Just curious to see if this works

Signed-off-by: Jesse Szwedko <jesse.szwedko@datadoghq.com>

jszwedko committed:
add make targets

Signed-off-by: Jesse Szwedko <jesse.szwedko@datadoghq.com>

wtaylor committed:
add image overrides for new arm v6 targets

protoc was previously not found due to `cross` falling back to the default build images instead of the image overlayed with protoc

Remove references to debs and rpms

wtaylor commented:
Was going to add CI targets for .rpm and .deb but there's a small conflict that I'm not knowledgeable enough to figure out.

As far as I understand it, Debian derived distros understand `armhf` as "Arm v7 with Hardware Floating Point registry". This would conflict with the new target of `armhf` as "Arm v6 with a Hardware Floating Point registry". As such I thought it'd be acceptable at least to get the started with just publishing to S3 and Docker where the triple is fully specified.

Salathaniel commented:
https://gptzero.me/?via=salathaniel

On Sun, Nov 19, 2023, 8:50 AM William Taylor ***@***.***>
wrote:

> Add ARMv6 builds using the work started in PR #18514
> <https://github.com/vectordotdev/vector/pull/18514>.
>
> Extends CI pipeline for the new target architecture to publish to docker
> and s3.
>
> fixes #18445 <https://github.com/vectordotdev/vector/issues/18445>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/vectordotdev/vector/pull/19192
> Commit Summary
>
>    - 06447a5
>    <https://github.com/vectordotdev/vector/pull/19192/commits/06447a5a30c1bae5e49c24533f19aa112db458ce>
>    feat(platforms): Add ARMv6 builds
>    - 2703c1f
>    <https://github.com/vectordotdev/vector/pull/19192/commits/2703c1f212620bf96ac5f2536e18995850ad6f23>
>    add make targets
>    - 6f1ca66
>    <https://github.com/vectordotdev/vector/pull/19192/commits/6f1ca6696bd9f1b9317707be310aacbb6d392533>
>    add image overrides for new arm v6 targets
>
> File Changes
>
> (7 files <https://github.com/vectordotdev/vector/pull/19192/files>)
>
>    - *M* .github/workflows/cross.yml
>    <https://github.com/vectordotdev/vector/pull/19192/files#diff-d1293e9cb6c8ff5bf3191c0f640a4f022a3d7c6d29f3edb56b545a8f93b783f9>
>    (2)
>    - *M* .github/workflows/publish.yml
>    <https://github.com/vectordotdev/vector/pull/19192/files#diff-551d1fcf87f78cc3bc18a7b332a4dc5d8773a512062df881c5aba28a6f5c48d7>
>    (90)
>    - *M* Cargo.toml
>    <https://github.com/vectordotdev/vector/pull/19192/files#diff-2e9d962a08321605940b5a657135052fbcef87b5e360662bb527c96d9a615542>
>    (8)
>    - *M* Cross.toml
>    <https://github.com/vectordotdev/vector/pull/19192/files#diff-3abca7e5bab5b140e44643dade038033b8a6f6e50771f76c7d48e7bd0b338815>
>    (6)
>    - *M* Makefile
>    <https://github.com/vectordotdev/vector/pull/19192/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52>
>    (19)
>    - *A* scripts/cross/arm-unknown-linux-gnueabihf.dockerfile
>    <https://github.com/vectordotdev/vector/pull/19192/files#diff-36253655f95da372332cb4f328742358d4783361641ad8d29f086e73e2077624>
>    (4)
>    - *A* scripts/cross/arm-unknown-linux-musleabihf.dockerfile
>    <https://github.com/vectordotdev/vector/pull/19192/files#diff-cab21e7bcf707ab9bae461b9e5eb34d7d7ce527ce9dc969277e334afad799f8b>
>    (8)
>
> Patch Links:
>
>    - https://github.com/vectordotdev/vector/pull/19192.patch
>    - https://github.com/vectordotdev/vector/pull/19192.diff
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/vectordotdev/vector/pull/19192>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARRFX3SZZBJ744GPGPKJ72DYFIFCLAVCNFSM6AAAAAA7RYDHZOVHI2DSMVQWIX3LMV43ASLTON2WKOZSGAYDAOBVHE3DGNY>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

wtaylor commented:
This still needs work, docker image fails to build due to not getting the artifact in with the expected name, currently investigating.

wtaylor commented:
:wave: This is now ready for review, I decided to forgo `hf` aka hardware float in the target, this will have a small impact on the efficiency on raspberry pi's but to the benefit of compatibility, this binary will work on all arm v6 devices. A Debian archive is now also built but I skipped creating an RPM, I decided against it as, as far as I can tell, there are no rpm based distros that support arm v6.

jszwedko commented:
Awesome, this is fantastic @wtaylor ! I'll give it a closer review soon.

jszwedko commented on .github/workflows/publish.yml:
```diff
@@ -207,6 +207,60 @@ jobs:
 
+  build-arm-unknown-linux-gnueabi-packages:
+    name: Build Vector for arm-unknown-linux-gnueabi (.tar.gz)
+    runs-on: [ linux, release-builder ]
```
```suggestion
    runs-on: [ linux, release-builder ]
    timeout-minutes: 60
```

We started adding these job timeouts proactively to avoid jobs inadvertently hanging for 6 hours by default. Is 60 minutes sufficient do you think? That's the timeout for the armv7 builds.

jszwedko reviewed (commented):
Thanks @wtaylor ! This is looks great. I'm happy to see both the glibc and musl-based builds. I just left one comment.

Out of curiosity: what issues did you run into trying to create the distroless docker images?

We'll also need to update the docs, but we can do that as a second pass after this is merged.

jszwedko commented:
I kicked off the workflow just to test: https://github.com/vectordotdev/vector/actions/runs/8112047694. This caused me to notice a couple of other things:

- I think the "generate checksums" job needs to depend on the added armv7 package builds (this job was added after you started this PR)
- For some reason, I thought this build was building debs (and not rpms) but it looks like not. I think that's OK if we just publish tarballs for the time being. We can add deb (and maybe rpm) package builds later

wtaylor commented:
> I kicked off the workflow just to test: https://github.com/vectordotdev/vector/actions/runs/8112047694. This caused me to notice a couple of other things:
> 
>     * I think the "generate checksums" job needs to depend on the added armv7 package builds (this job was added after you started this PR)
> 
>     * For some reason, I thought this build was building debs (and not rpms) but it looks like not. I think that's OK if we just publish tarballs for the time being. We can add deb (and maybe rpm) package builds later

I'll add the generate checksums job momentarily, it does actually build a deb for glibc, it's used in the debian based docker image, am I missing a step to publish it or something?

jszwedko commented:
Ah, also this needs a changelog entry 😄 See https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md

jszwedko commented:
> > I kicked off the workflow just to test: https://github.com/vectordotdev/vector/actions/runs/8112047694. This caused me to notice a couple of other things:
> > ```
> > * I think the "generate checksums" job needs to depend on the added armv7 package builds (this job was added after you started this PR)
> > 
> > * For some reason, I thought this build was building debs (and not rpms) but it looks like not. I think that's OK if we just publish tarballs for the time being. We can add deb (and maybe rpm) package builds later
> > ```
> 
> I'll add the generate checksums job momentarily, it does actually build a deb for glibc, it's used in the debian based docker image, am I missing a step to publish it or something?

Ah, you are right. I think what was confusing me is that the "Verify DEB" job only verifies the x86_64 deb packages and not armv6 (or armv7 or aarch).

wtaylor commented:
> Thanks @wtaylor ! This is looks great. I'm happy to see both the glibc and musl-based builds. I just left one comment.
> 
> Out of curiosity: what issues did you run into trying to create the distroless docker images?
> 
> We'll also need to update the docs, but we can do that as a second pass after this is merged.

Ah yeah so the core problem is simply that there isn't arm v6 builds of [distroless](https://github.com/GoogleContainerTools/distroless), they only build v7 upwards. We _could_ circumvent that by building the base images too but you'd then need a seperate pipeline and maintainence for that and it seems more hassle than it's worth, people with arm v6 will be more than happy to use debian or alpine based imaged :raised_hand_with_fingers_splayed: myself included.

wtaylor commented:
@jszwedko I've updated my forked branch and rebased onto the latest main whilst I was at it but my changes aren't here, you may need to convert to draft and convert back, I'll stick to linear additions from now on, my mistake :sweat_smile:

jszwedko commented:
No problem! Just let me know when it's ready for another round of review.

I noticed that one of the jobs fails:

```
2024-03-01T14:14:52.1956444Z  > [linux/amd64 builder 4/5] RUN ARCH=$(if [ "linux/amd64" = "linux/arm/v6" ]; then echo "arm"; else cat /etc/apk/arch; fi)     tar -xvf vector-0*-"$ARCH"-unknown-linux-musl*.tar.gz --strip-components=2:
2024-03-01T14:14:52.1957866Z 0.120 tar: can't open 'vector-0*--unknown-linux-musl*.tar.gz': No such file or directory
2024-03-01T14:14:52.1958688Z ------
2024-03-01T14:14:52.1988779Z Dockerfile:10
2024-03-01T14:14:52.2002836Z --------------------
2024-03-01T14:14:52.2003767Z    9 |     # special case for arm v6 builds, /etc/apk/arch reports armhf which conflicts with the armv7 package
2024-03-01T14:14:52.2005089Z   10 | >>> RUN ARCH=$(if [ "$TARGETPLATFORM" = "linux/arm/v6" ]; then echo "arm"; else cat /etc/apk/arch; fi) \
2024-03-01T14:14:52.2012184Z   11 | >>>     tar -xvf vector-0*-"$ARCH"-unknown-linux-musl*.tar.gz --strip-components=2
2024-03-01T14:14:52.2012791Z   12 |     
2024-03-01T14:14:52.2013076Z --------------------
2024-03-01T14:14:52.2015439Z ERROR: failed to solve: process "/bin/sh -c ARCH=$(if [ \"$TARGETPLATFORM\" = \"linux/arm/v6\" ]; then echo \"arm\"; else cat /etc/apk/arch; fi)     tar -xvf vector-0*-\"$ARCH\"-unknown-linux-musl*.tar.gz --strip-components=2" did not complete successfully: exit code: 1
2024-03-01T14:14:52.2018370Z Error: command: cd "/home/runner/work/vector/vector" && "/home/runner/work/vector/vector/scripts/build-docker.sh"
2024-03-01T14:14:52.2019067Z   failed with exit code: 1
2024-03-01T14:14:52.2019486Z make: *** [Makefile:626: release-docker] Error 1
2024-03-01T14:14:52.2035906Z ##[error]Process completed with exit code 2.
```

I attached the full log here since I'm not sure if outside contributors can see it in GHA.

[job-log.txt](https://github.com/vectordotdev/vector/files/14462719/job-log.txt)

wtaylor commented:
Ah it looks like [a partial outage](https://www.githubstatus.com/) is the reason that the PR is not synchronized, not my rebasing :D. May as well keep as a draft until I fix that expression anyway.

wtaylor committed:
Seperate arch var setting in alpine dockerfile due to subshell

wtaylor commented:
Amended and tested the alpine Dockerfile locally, to be honest, not a unix shell expert and not sure why separating the commands works I can only speculate that it's something to do with subshell execution. Anyway, ready for review again

jszwedko commented:
Thanks @wtaylor ! It got farther this time, but ran into a different error:

```
2024-03-04T16:54:37.5466028Z + build distroless-static 0.37.0.custom.8e2b116
2024-03-04T16:54:37.5466873Z + local BASE=distroless-static
2024-03-04T16:54:37.5467458Z + local VERSION=0.37.0.custom.8e2b116
2024-03-04T16:54:37.5468700Z + local TAG=timberio/vector:0.37.0.custom.8e2b116-distroless-static
2024-03-04T16:54:37.5469885Z + local DOCKERFILE=distribution/docker/distroless-static/Dockerfile
2024-03-04T16:54:37.5474221Z + '[' -n linux/amd64,linux/arm64,linux/arm/v7,linux/arm/v6 ']'
2024-03-04T16:54:37.5474996Z + ARGS=()
2024-03-04T16:54:37.5475443Z + [[ true == \t\r\u\e ]]
2024-03-04T16:54:37.5476071Z + ARGS+=(--push)
2024-03-04T16:54:37.5476870Z ++ evaluate_supported_platforms_for_base distroless-static
2024-03-04T16:54:37.5477842Z ++ local BASE=distroless-static
2024-03-04T16:54:37.5478420Z ++ IFS=,
2024-03-04T16:54:37.5478931Z ++ read -ra SUPPORTED_PLATFORMS_FOR_BASE
2024-03-04T16:54:37.5540412Z ++ local BUILDABLE_PLATFORMS=
2024-03-04T16:54:37.5541106Z ++ for platform in "${REQUESTED_PLATFORMS[@]}"
2024-03-04T16:54:37.5542354Z ++ [[ linux/amd64 linux/arm/v7 linux/arm64/v8 =~ linux/amd64 ]]
2024-03-04T16:54:37.5543288Z ++ BUILDABLE_PLATFORMS+=linux/amd64,
2024-03-04T16:54:37.5546667Z ++ for platform in "${REQUESTED_PLATFORMS[@]}"
2024-03-04T16:54:37.5547679Z ++ [[ linux/amd64 linux/arm/v7 linux/arm64/v8 =~ linux/arm64 ]]
2024-03-04T16:54:37.5548589Z ++ BUILDABLE_PLATFORMS+=linux/arm64,
2024-03-04T16:54:37.5549369Z ++ for platform in "${REQUESTED_PLATFORMS[@]}"
2024-03-04T16:54:37.5550648Z ++ [[ linux/amd64 linux/arm/v7 linux/arm64/v8 =~ linux/arm/v7 ]]
2024-03-04T16:54:37.5551512Z ++ BUILDABLE_PLATFORMS+=linux/arm/v7,
2024-03-04T16:54:37.5552290Z ++ for platform in "${REQUESTED_PLATFORMS[@]}"
2024-03-04T16:54:37.5553179Z ++ [[ linux/amd64 linux/arm/v7 linux/arm64/v8 =~ linux/arm/v6 ]]
2024-03-04T16:54:37.5554741Z ++ echo 'WARN: skipping linux/arm/v6 for distroless-static, no base image for platform'
2024-03-04T16:54:37.5556242Z WARN: skipping linux/arm/v6 for distroless-static, no base image for platform
2024-03-04T16:54:37.5557225Z ++ echo linux/amd64,linux/arm64,linux/arm/v7
2024-03-04T16:54:37.5558075Z + local BUILDABLE_PLATFORMS=linux/amd64,linux/arm64,linux/arm/v7
2024-03-04T16:54:37.5560667Z + docker buildx build --platform=linux/amd64,linux/arm64,linux/arm/v7,linux/arm/v6 --tag timberio/vector:0.37.0.custom.8e2b116-distroless-static target/artifacts -f distribution/docker/distroless-static/Dockerfile --push
2024-03-04T16:54:37.9137919Z #0 building with "builder-5f88644b-d7cb-41ac-bd1b-b27fa84e369d" instance using docker-container driver
2024-03-04T16:54:37.9139102Z 
2024-03-04T16:54:37.9139374Z #1 [internal] load build definition from Dockerfile
2024-03-04T16:54:37.9139945Z #1 transferring dockerfile: 617B done
2024-03-04T16:54:37.9140699Z #1 DONE 0.0s
2024-03-04T16:54:37.9140909Z 
2024-03-04T16:54:37.9141255Z #2 [linux/amd64 internal] load metadata for docker.io/library/alpine:3.18
2024-03-04T16:54:37.9141893Z #2 DONE 0.1s
2024-03-04T16:54:37.9142085Z 
2024-03-04T16:54:37.9142432Z #3 [linux/arm/v6 internal] load metadata for docker.io/library/alpine:3.18
2024-03-04T16:54:37.9143060Z #3 DONE 0.1s
2024-03-04T16:54:37.9143247Z 
2024-03-04T16:54:37.9143516Z #4 [linux/arm64 internal] load metadata for docker.io/library/alpine:3.18
2024-03-04T16:54:37.9144039Z #4 DONE 0.1s
2024-03-04T16:54:37.9144192Z 
2024-03-04T16:54:37.9144477Z #5 [linux/arm/v7 internal] load metadata for docker.io/library/alpine:3.18
2024-03-04T16:54:37.9144992Z #5 DONE 0.1s
2024-03-04T16:54:37.9145139Z 
2024-03-04T16:54:37.9145434Z #6 [linux/amd64 internal] load metadata for gcr.io/distroless/static:latest
2024-03-04T16:54:37.9778743Z #6 ...
2024-03-04T16:54:37.9779399Z 
2024-03-04T16:54:37.9780245Z #7 [linux/arm/v6 internal] load metadata for gcr.io/distroless/static:latest
2024-03-04T16:54:37.9781579Z #7 ERROR: no match for platform in manifest: not found
2024-03-04T16:54:37.9866351Z 
2024-03-04T16:54:37.9866891Z #6 [linux/amd64 internal] load metadata for gcr.io/distroless/static:latest
2024-03-04T16:54:37.9867473Z #6 CANCELED
2024-03-04T16:54:37.9867637Z 
2024-03-04T16:54:37.9867996Z #8 [linux/arm/v7 internal] load metadata for gcr.io/distroless/static:latest
2024-03-04T16:54:37.9868561Z #8 CANCELED
2024-03-04T16:54:37.9868717Z 
2024-03-04T16:54:37.9868981Z #9 [linux/arm64 internal] load metadata for gcr.io/distroless/static:latest
2024-03-04T16:54:37.9870624Z #9 CANCELED
2024-03-04T16:54:37.9871205Z ------
2024-03-04T16:54:37.9871894Z  > [linux/arm/v6 internal] load metadata for gcr.io/distroless/static:latest:
2024-03-04T16:54:37.9872781Z ------
2024-03-04T16:54:37.9883477Z Dockerfile:12
2024-03-04T16:54:37.9884185Z --------------------
2024-03-04T16:54:37.9884829Z   10 |     # distroless doesn't use static tags
2024-03-04T16:54:37.9885548Z   11 |     # hadolint ignore=DL3007
2024-03-04T16:54:37.9886073Z   12 | >>> FROM gcr.io/distroless/static:latest
2024-03-04T16:54:37.9886460Z   13 |     
2024-03-04T16:54:37.9886880Z   14 |     COPY --from=builder /vector/bin/* /usr/local/bin/
2024-03-04T16:54:37.9887322Z --------------------
2024-03-04T16:54:37.9887885Z ERROR: failed to solve: gcr.io/distroless/static:latest: no match for platform in manifest: not found
2024-03-04T16:54:37.9909500Z Error: command: cd "/home/runner/work/vector/vector" && "/home/runner/work/vector/vector/scripts/build-docker.sh"
2024-03-04T16:54:37.9910767Z   failed with exit code: 1
2024-03-04T16:54:37.9919532Z make: *** [Makefile:626: release-docker] Error 1
```

Full log:  [job-logs.txt](https://github.com/vectordotdev/vector/files/14485029/job-logs.txt)

wtaylor committed:
use filtered platforms for buildx

wtaylor commented:
> Thanks @wtaylor ! It got farther this time, but ran into a different error:

The bane of PR's to CI systems, fixed

jszwedko commented:
> > Thanks @wtaylor ! It got farther this time, but ran into a different error:
> 
> The bane of PR's to CI systems, fixed

For sure 😓 I'll kick off another build.

jszwedko commented:
It looks like the custom build workflow jobs all passed 🎉 I'll give this another look shortly.

It looks like there were some shell lints flagged if you get some time.

wtaylor committed:
fix shellcheck errors

wtaylor commented:
> It looks like the custom build workflow jobs all passed 🎉 I'll give this another look shortly.
> 
> It looks like there were some shell lints flagged if you get some time.

A new tool on the tool belt for me, all shellcheck issues fixed

jszwedko reviewed (approved):
Thanks for this contribution @wtaylor !
